### PR TITLE
Feature/handle compression

### PR DESCRIPTION
**Pull request by kevin3314**
Handle compression file.

There are two kind of usage: archive file (e.g. zip) and compressed file (e.g. gzip).
Compressed file in archive file is not allowed because there seems no way to open it.

One challenge in parsing archive file is that unarchive is incompatible with multiprocessing:
Unarchived instance (i.e. ZipFile) is not picklable and can not be passed to multiprocessing, so we need to unarchive every time a file is opened.
It is time-consuming so I decide that multiprocessing is disabled when dealing with an archive file.

Generic use case is covered by test, but there are some cases which is not covered. Specifically, pickle file (w/ w/o gzip) is not tested.

### CODE REVIEW

**Comment by nobu-g** [comment]
Thanks a lot for your contributions, such as comprehensive support for compressed and archived files, clear explanations for the intention and the limitation, and even adding tests!

BTW, I have an idea for more readable code; How about delegating archive-related logic to `ArchiveHandler` class more? Specifically, I want to implement methods such as `list_contents` and `open`, which abstracts `getnames` and `namelist` etc... This enables the `KyotoReader.__init__` method to have a logic that is less aware of input format.

If you do not have enough time, I'll try this :)

File: src/kyoto_reader/reader.py [hunk 1]
```diff
@@ -49,21 +76,55 @@ def __init__(self,
                  ) -> None:
         if not (isinstance(source, Path) or isinstance(source, str)):
             raise TypeError(f"document source must be Path or str type, but got '{type(source)}' type")
+        # Yields all allowed single-file extension (e.g. .knp, .pkl.gz)
+        ALLOWED_SINGLE_FILE_EXT = list(
```
Comment (nobu-g) [inline src/kyoto_reader/reader.py:80]: Variable in function should be lowercase

File: src/kyoto_reader/reader.py [hunk 2]
```diff
@@ -49,21 +76,55 @@ def __init__(self,
...
+            logger.info("Multiprocessing with archive is too slow, so it is disabled")
+            logger.info(
+                "Run without multiprocessing can be relatively slow, so please consider unarchive the archive file")
+            self.mp_backend = None
         self.n_jobs: int = n_jobs
 
-        args_iter = ((path, did_from_sid) for path in file_paths if path.suffix == knp_ext)
-        rets: List[Dict[str, str]] = self._mp_wrapper(KyotoReader.read_knp, args_iter, self.mp_backend, self.n_jobs)
+        if self.archive_path is not None:
+            args_iter = (
```
Comment (nobu-g) [inline src/kyoto_reader/reader.py:114]: This variable is not used

File: src/kyoto_reader/reader.py [hunk 3]
```diff
@@ -110,6 +177,21 @@ def read_knp(path: Path, did_from_sid: bool) -> Dict[str, str]:
...
+            with archive.open(str(path)) as f:
+                text = f.read().decode("utf-8")
+                _read_knp(text.split("\n"))
+        else:
+            if any(key in path.suffixes for key in KyotoReader.COMPRESS2OPENER):
+                compress = set(path.suffixes) & set(KyotoReader.COMPRESS2OPENER.keys())
+                assert len(compress) == 1
+                opener = KyotoReader.COMPRESS2OPENER[compress.pop()]
+            else:
+                opener = open
```
Comment (nobu-g) [inline src/kyoto_reader/reader.py:191]: A bit strange substitution? ("opener" is a kind of actor, and "open" is a kind of action)

File: tests/test_document.py [hunk 4]
```diff
@@ -400,3 +401,28 @@ def test_ne(fixture_kyoto_reader: KyotoReader):
...
+    data_dir = Path(__file__).parent / 'data'
+    gzip_reader = KyotoReader(data_dir / 'gzip_knp',
+                         target_cases=ALL_CASES,
+                         target_corefs=ALL_COREFS,
+                         mp_backend=None)
+    assert fixture_kyoto_reader.doc_ids == gzip_reader.doc_ids
+    assert all(
+        fixture_kyoto_reader.process_document(doc_id) == gzip_reader.process_document(doc_id)
+        for doc_id in fixture_kyoto_reader.doc_ids
+    )
```
Comment (nobu-g) [inline tests/test_document.py:428]: No new line at the end of file

**Comment by kevin3314** [comment]
> BTW, I have an idea for more readable code;
I think that would be great. I'll implement them in a few days! (If that does not come out, please implement 🙇‍♂️)

**Pull request merged**